src/Shared/runtime/Http2/Hpack/HPackDecoder.cs

```diff
@@ -577,7 +577,7 @@ private void OnStringLength(int length, State nextState)
                     throw new HPackDecodingException(SR.Format(SR.net_http_headers_exceeded_length, _maxHeadersLength));
                 }
 
-                _stringOctets = new byte[Math.Max(length, _stringOctets.Length * 2)];
+                _stringOctets = new byte[Math.Max(length, Math.Min(_stringOctets.Length * 2, _maxHeadersLength))];
```

Comment:
There is a `EnsureStringCapacity` method in this file that also doubles an array size. Does it need to be changed?